[WIP] Benchpark New Developer Docs (#786)
LLNL/benchpark: docs/developer-docs/benchpark-developer-docs.rst

===============================================
Benchpark Design Concepts and Command Workflow
===============================================

> This document outlines the important concepts and patterns used consistently in Benchpark (BP) code design, along with detailed explanations of key commands and their workflow.

Reviewer: We should always enumerate `Benchpark` and not use shorthand `BP`

Table of Contents
=================

This document covers the following main sections:

* `Core Concepts and Patterns`_ - Understanding Variants, Consistent Flow, and Modifiers
* `Benchpark Command Workflow`_ - Step-by-step execution of key Benchpark commands  
* `Debugging Benchpark with VSCode`_ - Setting up debugging environment for development

Core Concepts and Patterns
==========================

1. Variants
-----------

> Variants are predefined keywords that execute specific logic and set configurations for benchmarks based on whether these keywords are found in user arguments.

Reviewer: ```suggestion
Variants are directives that execute specific logic and set configurations for experiments, systems, and packages. Variants are of the form `name=value`, where the `value` is selected from a set of possible values, defined by the given experiment, system, or package.
```

* ``self`` contains all the Specs, in other words the extra arguments passed in the user command, defined for the current context
* ``self.name`` contains the name of the core component spec relative to the current context, examples:
  
  * ``llnl-cluster`` in benchmark system commands running BP on LLNL systems
    
    * At executing benchpark system command (``benchpark system init --dest=.... llnl-cluster cluster=...``), ``self.name`` resolves to ``llnl-cluster`` since it is the core component, as it defines the system we will run our experiments on
  
  * ``saxpy`` when running benchmark experiment command
    
    * At executing benchpark experiment command (``benchpark experiment init --dest=... saxpy``), ``self.name`` resolves to ``saxpy`` as it is the core component that defines the benchmark that will be used
* A valid question could be, why not the ``cluster=ruby`` be used as ``self.name``?

  * Based on my experiments, order of args passed to benchpark does really matters. By default our parser expects key-value pairs to be at the end, and prior is the Spec name.

* We get the parent directory of the Python class path responsible for creating the ``self.name`` object

> * ``self.name`` is set when creating the corresponding ChildSpec object, (``SystemSpec``, ``ExperimentSpec``) where they inherit from ``Spec`` class (the parent class). This is done right before executing ``concretize()``

Reviewer: ```suggestion
* ``self.name`` is set when creating the corresponding child object, (``SystemSpec``, ``ExperimentSpec``) where they inherit from ``Spec`` class (the parent class). This is done right before executing ``concretize()``
```

* We depend on Ramble for resolving this

**Example:**

For the command::

    benchpark system init --dest=test-ruby-system llnl-cluster cluster=ruby

* ``self.name`` will be resolved to ``llnl-cluster``
* This resolves to get the path of the ``LlnlCluster`` class that contains the variants it accepts

**Variant Processing:**

1. Ramble gets all variants (and dependent variants from any imports of the class) and registers them
2. We check user-provided variants against all acceptable variants to ensure they are as expected
3. Passed user variants are added to generated YAML files
4. Later, we import needed modifiers or load required libraries/tools to provide the requested functionality

2. Consistent Flow for Benchpark Commands
-----------------------------------------

**Flow Overview:**

1. **Entry Point:** ``benchpark`` → ``lib/main.py``
   
   * Gets needed imports, including ``cmd.experiment``
   * Executes ``benchpark.experiment`` (``lib/benchpark/experiment.py``)
   * Checks if Ramble and Spack exist in home directory

>    * Clones and installs them if they don't exist

Reviewer: ```suggestion
   * Clones Ramble and Spack libraries in the home directory (`~/.benchpark`) if they do not exist.
```

2. **Command Processing:**
   
   * Lists all acceptable commands
   * Parses user command to determine which action to execute (e.g., experiment, system)
   * Executes the corresponding command function:
     
     * ``experiment.py`` → ``command()``
     * ``system.py`` → ``command()``
     * etc.

3. Modifiers
------------

Modifiers provide extra functionality to:

* Help gather more information (such as affinity data)
* Enable tools (such as Caliper) to produce ``.cali`` files for later analysis using Thicket

Benchpark Command Workflow
==========================

> 1. System Initialization

Reviewer: For these sections, we would probably need to explain in terms of the functions that are used to create the yamls, e.g. in `system.py`:
```
    def compute_dict(self):
        # This can be overridden by any subclass that needs more flexibility
        return {
            "system_id": self.compute_system_id(),
            "variables": self.compute_variables_section(),
            "software": self.compute_software_section(),
            "auxiliary_software_files": {
                "compilers": self.compute_compilers_section(),
                "packages": self.compute_packages_section(),
            },
        }
```

------------------------

**Command:**

.. code-block:: bash

   benchpark system init --dest=test-ruby-system llnl-cluster cluster=ruby

**Generated Output:**

* **YAML files:** Define software needed (to be installed by Spack later)

> * **YAML files:** Define the LLNL cluster system to be used

Reviewer: Would probably need to define what each is for `auxiliary_software_files/compilers.yaml` compilers specified by benchpark, `auxiliary_software_files/packages.yaml` external packages for spack, `software.yaml`, `system_id.yaml`, `variables.yaml` 

edit: I see one of these below

**Generated Directory Structure:** ``test-ruby-system/``

* ``variables.yaml``: Contains configurations for job execution (#nodes, #cores, etc.)

2. Experiment Initialization
----------------------------

**Command:**

.. code-block:: bash

   benchpark experiment init --dest=test-amg2023-benchmark amg2023 +openmp

**Generated Directory Structure:** ``test-amg2023-benchmark/``

* ``ramble.yaml``: 
  

>   * Defines problems and experiments to run

Reviewer: ```suggestion
  * Defines experiment variables, which ramble uses to set up experiments.
```

  * Contains required modifiers and packages to be installed

> 3. Setup Workspace

Reviewer: ```suggestion
3. Setup Benchpark Workspace
```

------------------

**Command:**

.. code-block:: bash

   benchpark setup ./test-amg2023-benchmark ./test-ruby-system test-workspace/

**Generated Directory Structure:** ``test-workspace/``

**New Directory Created:** ``test-amg2023-benchmark/test-ruby-system/workspace``

* Based on definitions in ``test-ruby-system`` and ``test-amg2023-benchmark``
* Contains everything as defined before this step

> * Creates necessary scripts for configuring Ramble and Spack installation

Reviewer: ```suggestion
* Clones Ramble and Spack into the benchpark workspace
```

4. Environment Setup
--------------------

**Command:**

.. code-block:: bash

   . test-workspace/setup.sh

**Purpose:**

> * Runs two simple scripts to set up Ramble and Spack previously installed for next steps

Reviewer: This disables the local spack configuration, so a users external spack packages will not be registered by the spack in the benchpark workspace. The benchpark workspace Ramble and Spack are added to the users path so they can use `ramble` and `spack` commands.

5. Workspace Configuration
--------------------------

**Location:**

.. code-block:: bash

   cd ./test-workspace/test-amg2023-benchmark/test-ruby-system/workspace/

**Command:**

.. code-block:: bash

   ramble --workspace-dir . --disable-progress-bar workspace setup

**What Happens:**

* **Spack Role:** Installs all needed software

> * **Ramble Role:** Sets up all experiments/problems defined previously for execution

Reviewer: ```suggestion
* **Ramble Role:** Sets up all experiments/problems defined in the experiment `ramble.yaml`
```

* **Benchpark Role:** Complete at this point

**Generated Content:**

> * New directory ``experiments`` is created

Reviewer: ```suggestion
* New directory ``experiments`` is created for each experiment in the `ramble.yaml`
```

* Contains script ``execute_experiment`` for the next step

> * This script is the main one that Ramble runs to fire jobs for experiments on the HPC cluster

Reviewer: ```suggestion
* This is a job script submitted to the scheduler
```

* Script includes all needed steps to run experiments using:
  
  * Requested benchmark
  * Configurations
  * Packages
  * Libraries
  * etc.

6. Experiment Execution
-----------------------

**Command:**

.. code-block:: bash

   ramble --disable-progress-bar --workspace-dir . on

**Purpose:**

> * Runs the ``execute_experiment`` script generated from the previous step

Reviewer: ```suggestion
* Submits the ``execute_experiment`` job script generated from the previous step to the scheduler
```

> * Execution occurs on the chosen LLNL cluster (e.g., Ruby or Dane)

Reviewer: ```suggestion

```
